### PR TITLE
Use FHIR Patient resource to persist user info

### PR DESCRIPTION
**Pull request by namalu**
- Use FHIR Patient resource to persist user info globally.
- UsersService unit tests.

### CODE REVIEW

File: src/Common/FitOnFhir.Common/Repositories/TableRepository.cs [hunk 1]
```diff
@@ -38,11 +38,17 @@ public async Task<TEntity> GetById(string id, CancellationToken cancellationToke
...
+            {
+                var response = await _tableClient.GetEntityAsync<TableEntity>(PartitionKey, id, cancellationToken: cancellationToken);
 
-            if (response?.Value != null)
+                if (response?.Value != null)
+                {
+                    return (TEntity)Activator.CreateInstance(typeof(TEntity), response.Value);
+                }
+            }
+            catch (RequestFailedException e) when (e.Status == 404)
```
Comment (cgilbert76) [inline src/Common/FitOnFhir.Common/Repositories/TableRepository.cs:50]: Are we not doing anything in this catch block?

File: src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        private readonly ResourceManagementService _resourceManagementService;
+        private readonly IUsersTableRepository _usersTableRepository;
+
+        public UsersServiceBase(ResourceManagementService resourceManagementService, IUsersTableRepository usersTableRepository)
+        {
+            _resourceManagementService = EnsureArg.IsNotNull(resourceManagementService, nameof(resourceManagementService));
+            _usersTableRepository = EnsureArg.IsNotNull(usersTableRepository, nameof(usersTableRepository));
+        }
+
+        public async System.Threading.Tasks.Task EnsurePatientAndUser(string platformName, string identifier, string system, CancellationToken cancellationToken)
```
Comment (cgilbert76) [inline src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs:25]: Does this namespace need to be fully qualified?

File: src/GoogleFit/FitOnFhir.GoogleFit.Tests/UsersServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        protected override string ExpectedPlatform => GoogleFitConstants.GoogleFitPlatformName;
+
+        [Fact]
+        public async Task GivenAuthCodeIsNull_WhenProcessAuthorizationCallbackCalled_ConditionIsLogged()
+        {
+            await _usersService.ProcessAuthorizationCallback(null, CancellationToken.None);
+
+            _logger.Received(1).Log(
+                Arg.Is<LogLevel>(lvl => lvl == LogLevel.Information),
+                Arg.Is<string>(msg => msg == "ProcessAuthorizationCallback called with no auth code"));
```
Comment (cgilbert76) [inline src/GoogleFit/FitOnFhir.GoogleFit.Tests/UsersServiceTests.cs:66]: I've always been a little iffy on these tests that check for logger calls.  Logging can sometimes be very informal, and if we are testing for the exactness of the message logged itself, it seems quite possible that down the road someone may decide they want to change a logger line for debugging purposes in the moment.  In that case, tests like these will fail, when we shouldn't necessarily be holding a standard to the logging message itself.

File: src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            EnsureArg.IsNotNullOrWhiteSpace(system, nameof(system));
+
+            // Get or create a Patient resource.
+            Patient patient = await _resourceManagementService.EnsureResourceByIdentityAsync<Patient>(
+                identifier,
+                system,
+                (p, id) => p.Identifier = new List<Identifier> { id });
+
+            // Check to see if the user exists in the repository.
+            User user = await _usersTableRepository.GetById(patient.Id, cancellationToken);
```
Comment (cgilbert76) [inline src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs:38]: Wondering what happens, the way this is written, if a user exists already but with a different platform saved, then won't this skip saving the google fit platform info below?

**Review by cgilbert76 — CHANGES_REQUESTED** [review]

File: src/Common/FitOnFhir.Common/Repositories/TableRepository.cs (hunk 1 above)
Comment (namalu) [reply src/Common/FitOnFhir.Common/Repositories/TableRepository.cs:50]: That is correct. The only exception that we catch here is a 404 not found, and we then return null. the behavior before catching a not found was execution would halt or the caller needed to handle the 404.

File: src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs (hunk 2 above)
Comment (namalu) [reply src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs:25]: Unfortunately, yes. Hl7.Fhir.Model has a type named Task. It's frustrating.

File: src/GoogleFit/FitOnFhir.GoogleFit.Tests/UsersServiceTests.cs (hunk 3 above)
Comment (namalu) [reply src/GoogleFit/FitOnFhir.GoogleFit.Tests/UsersServiceTests.cs:66]: That is true. However, I did want some definitive way to test that this condition was actually handled appropriately. If the message is changed, it's probably a good idea that the test also needs to be updated.

File: src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs (hunk 4 above)
Comment (namalu) [reply src/Common/FitOnFhir.Common/Services/UsersServiceBase.cs:38]: Good catch. We should talk this through on Monday. Basically, we need to have an external identifier to resolve this problem. When a user starts the authorization flow, they might need to pass in an identifier (or a token oid once the method is no longer anonymous) that we can use to link all platforms to a single person.

**Review by cgilbert76 — APPROVED** [review]

**Pull request merged**